### PR TITLE
exp/clients/horizon: Fetch timebounds

### LINKED ISSUE #1157: exp/clients/horizon: fetch Horizon server time
We have Timebounds support in `txnbuild` (in progress), including setting a timeout from system clock.

In [this commit](https://github.com/stellar/js-stellar-sdk/commit/6cc54eb14d92671f7e00e846e8f8dcd7d7f7c4bd), the JS SDK adds support for [retrieving the server time](https://stellar.github.io/js-stellar-sdk/Server.html#fetchTimebounds) of a Horizon instance, to be used in constructing timebounds. The idea is that the server time may be a more correct clock to compare against if making a timebound based on a timeout.

This issue is to consider and possibly implement something similar for the Go SDK.

### PR DESCRIPTION
**Pull request by poliha**
This PR adds support for creating transaction timebounds using the server time of an horizon client.
It defaults to using the local time when no server time has been saved. 
Typically, server time is saved on every request to the server, so when you request the details of a source account before building a transaction, the current server time will be saved. 
It closes #1157

### CODE REVIEW

File: exp/clients/horizon/client.go [hunk 1]
```diff
@@ -462,5 +464,23 @@ func (c *Client) StreamOrderBooks(ctx context.Context, request OrderBookRequest,
...
+	serverURL, err := url.Parse(c.HorizonURL)
+	if err != nil {
+		panic("Unable to parse horizon url")
+	}
+	currentTime := currentServerTime(serverURL.Hostname())
+	if currentTime != 0 {
+		return txnbuild.NewTimebounds(0, currentTime+seconds)
+	}
+
+	// return local time if no server time has been recorded
```
Comment (ire-and-curses) [inline exp/clients/horizon/client.go]: Comment is a little misleading. Could say `return a timebounds based on local time if no server time has been recorded`

File: exp/clients/horizon/client.go [hunk 2]
```diff
@@ -462,5 +464,23 @@ func (c *Client) StreamOrderBooks(ctx context.Context, request OrderBookRequest,
...
+	// to do: query an endpoint to get the most current time. Implement this after we add retry logic to client.
+	return txnbuild.NewTimebounds(0, time.Now().Unix()+seconds)
```
Comment (ire-and-curses) [inline exp/clients/horizon/client.go]: In this case I think you should use the other builder method: `return txnbuild.NewTimeout(seconds)`. It's simpler and keeps the logic in one place (also here you forgot the UTC()).

File: exp/clients/horizon/internal.go [hunk 3]
```diff
@@ -13,6 +14,11 @@ func decodeResponse(resp *http.Response, object interface{}) (err error) {
 	defer resp.Body.Close()
 	decoder := json.NewDecoder(resp.Body)
 
+	// resp.Request should not be nil for Client requests
+	if resp.Request != nil {
```
Comment (ire-and-curses) [inline exp/clients/horizon/internal.go]: Should we return an error if somehow it is `nil`?

File: exp/clients/horizon/internal.go [hunk 4]
```diff
@@ -99,3 +105,26 @@ func addQueryParams(params ...interface{}) string {
 
 	return query.Encode()
 }
+
+func setCurrentServerTime(host string, serverDate []string) {
```
Comment (ire-and-curses) [inline exp/clients/horizon/internal.go]: Add docstring

File: exp/clients/horizon/internal.go [hunk 5]
```diff
@@ -99,3 +105,26 @@ func addQueryParams(params ...interface{}) string {
...
+	st, err := time.Parse(time.RFC1123, serverDate[0])
+	if err != nil {
+		return
+	}
+	ServerTimeMap[host] = ServerTimeRecord{ServerTime: st.Unix(), LocalTimeRecorded: time.Now().Unix()}
```
Comment (ire-and-curses) [inline exp/clients/horizon/internal.go]: - `ServerTime` should be `st.UTC().Unix()`
- `LocalTimeRecorded` should be `time.Now().UTC().Unix()`

File: exp/clients/horizon/internal.go [hunk 6]
```diff
@@ -99,3 +105,26 @@ func addQueryParams(params ...interface{}) string {
...
+}
+
+func currentServerTime(host string) int64 {
+	st := ServerTimeMap[host]
+	if &st == nil {
+		return 0
+	}
+
+	currentTime := time.Now().Unix()
```
Comment (ire-and-curses) [inline exp/clients/horizon/internal.go]: `currentTime:= time.Now().UTC().Unix()`

File: exp/clients/horizon/internal.go [hunk 7]
```diff
@@ -99,3 +105,26 @@ func addQueryParams(params ...interface{}) string {
...
+	// if it's been more than 5 minutes from the last time, then return 0
```
Comment (ire-and-curses) [inline exp/clients/horizon/internal.go]: this comment could indicate why - because the server time is way off

File: exp/clients/horizon/main.go [hunk 8]
```diff
@@ -295,3 +295,12 @@ type TradeAggregationRequest struct {
 	Order              Order
 	Limit              uint
 }
+
+// ServerTimeRecord contains data for the current unix time of an horizon server instance, and the local time when it was recorded
```
Comment (ire-and-curses) [inline exp/clients/horizon/main.go]: `an horizon server` -> `a horizon server`

File: exp/clients/horizon/main_test.go [hunk 9]
```diff
@@ -908,6 +911,26 @@ func TestOrderBookRequest(t *testing.T) {
...
+	client := DefaultPublicNetClient
+	_, err := client.Metrics()
+
+	st := client.FetchTimebounds(100)
+	if assert.NoError(t, err) {
+		assert.IsType(t, ServerTimeMap["horizon.stellar.org"], ServerTimeRecord{})
+		assert.Equal(t, st.MinTime, int64(0))
+	}
+
+	newRecord := ServerTimeRecord{ServerTime: 100, LocalTimeRecorded: time.Now().Unix()}
```
Comment (ire-and-curses) [inline exp/clients/horizon/main_test.go]: `time.Now().UTC().Unix()`

File: exp/clients/horizon/internal.go (hunk 3 above)
Comment (poliha) [reply exp/clients/horizon/internal.go]: When using mocks this is nil. So if we return error, the tests will fail.

File: exp/clients/horizon/client.go [hunk 10]
```diff
@@ -462,5 +464,23 @@ func (c *Client) StreamOrderBooks(ctx context.Context, request OrderBookRequest,
 	return request.StreamOrderBooks(ctx, c, handler)
 }
 
+// FetchTimebounds get timebounds for N seconds from now using the server time of an horizon instance.
+// It defaults to localtime when the server time is not available
+// Note that this will generate your timebounds when you init the transaction, not when you build or submit the transaction! So give yourself enough time to get the transaction built and signed before submitting.
+func (c *Client) FetchTimebounds(seconds int64) txnbuild.Timebounds {
+	serverURL, err := url.Parse(c.HorizonURL)
+	if err != nil {
+		panic("Unable to parse horizon url")
```
[the remark this answers is not in the document]
Comment (poliha) [reply exp/clients/horizon/client.go]: yea error should be fine here

**Pull request merged**